Feat/implement-refresh-tokens

laipz8200 opened the pull request:
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

- Add `ACCESS_TOKEN_EXPIRE_MINUTES` to `.env.example` files and `docker-compose.yaml` for setting access token validity.
- Update `get_account_jwt_token` to use the configured expiration time.
- Rename `OAuthConfig` to `AuthConfig` to encompass general authentication settings.
- Correct typographical error in GitHub OAuth configuration description.

Fixes 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

laipz8200 committed:
feat(auth): implement refresh tokens and enhance login process

- Introduced `TokenPair` model for managing access and refresh tokens.
- Added `refresh_token` method to generate new tokens upon expiration.
- Updated login/logout processes to handle token pairs and enhanced security.
- Replaced `get_remote_ip` with `extract_remote_ip` for clarity.
- Added endpoint for refreshing tokens to maintain user session continuity.

bowenliang123 commented on api/services/account_service.py:
```diff
@@ -84,10 +113,13 @@ def load_user(user_id: str) -> None | Account:
-            "exp": datetime.now(timezone.utc).replace(tzinfo=None) + exp,
+            "exp": int(
+                datetime.now(timezone.utc).timestamp()
+                + timedelta(minutes=dify_config.ACCESS_TOKEN_EXPIRE_MINUTES).total_seconds()
```
Considering the expiration time is used in seconds, it's suggested to change the time unit of this config from minutes to seconds for better consistency and flexibility.

laipz8200 replied:
Thank you for your suggestion! For this scenario, using minutes is sufficient and more readable, as there won't be any access tokens with a validity period of less than one minute.

bowenliang123 replied:
SGTM.

bowenliang123 commented on api/services/account_service.py:
```diff
@@ -698,3 +749,8 @@ def _get_invitation_by_token(cls, token: str, workspace_id: str, email: str) ->
             return invitation
+
+
+def _generate_refresh_token(length: int = 64):
```
```suggestion
def _generate_refresh_token(length: int = 64)->str:
```

laipz8200 replied:
Considering that in this situation the LSP has sufficient type inference capabilities, I personally lean towards not adding this annotation.

laipz8200 committed:
refactor(auth): simplify JWT token expiration

- Streamlined expiration time calculation by separating date arithmetic from payload construction.
- Improved code readability and maintainability.
